### PR TITLE
commands/build: separate planning from execution (CRAFT-331)

### PR DESCRIPTION
**Pull request by cjp256**
Determine bases that can be built first (in new `plan()` method)
then run through the process of building.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: charmcraft/commands/build.py [hunk 1]
```diff
@@ -300,10 +300,10 @@ def _set_prime_filter(self):
...
                 self._prime.append(fn)
 
-    def run(
-        self, bases_indices: Optional[List[int]] = None, destructive_mode: bool = False
-    ) -> List[str]:
-        """Run build process.
+    def plan(
+        self, *, bases_indices: Optional[List[int]], destructive_mode: bool, managed_mode: bool
+    ) -> List[Tuple[BasesConfiguration, Base, int, int]]:
+        """Plan build porocess.
```
Comment (cmatsuoka) [inline charmcraft/commands/build.py:306]: Typo here in process.

File: charmcraft/commands/build.py [hunk 2]
```diff
@@ -371,11 +349,62 @@ def run(
...
             raise CommandError(
                 "No suitable 'build-on' environment found in any 'bases' configuration."
             )
 
+        charms = []
+        for bases_config, build_on, bases_index, build_on_index in build_plan:
+            logger.debug(
+                "Building for 'bases[%d][%d]'.",
+                bases_index,
+                build_on_index,
```
Comment (cmatsuoka) [inline charmcraft/commands/build.py:396]: Maybe the arguments could be in a single line?

**Review by cmatsuoka — APPROVED** [review]
Looks good! (but there's a typo do fix)

File: charmcraft/commands/build.py [hunk 3]
```diff
@@ -312,20 +312,7 @@ def run(
 
         :returns: List of charm files created.
```
Comment (facundobatista) [inline charmcraft/commands/build.py:313]: This is not longer true, right?

**Review by facundobatista — APPROVED** [review]
Like it. Annotated a small detail.

File: charmcraft/commands/build.py (hunk 2 above)
Comment (cjp256) [reply charmcraft/commands/build.py]: done!

File: charmcraft/commands/build.py (hunk 1 above)
Comment (cjp256) [reply charmcraft/commands/build.py]: Fixed and improved the entire doc block

File: charmcraft/commands/build.py (hunk 3 above)
Comment (cjp256) [reply charmcraft/commands/build.py]: Indeed, I rewrote the entire doc string, thanks!

**Pull request merged**